content/en/docs/v3/guides/infra/tls_dns.md

```diff
@@ -0,0 +1,140 @@
+
+To start with we are focussed on GCP but will expand to other cloud providers.
+
+# Google Cloud Platform
+
+## Prerequisites
+
+- cluster created using Jenkins X [GCP Terraform getting started](/docs/v3/getting-started/gke/)
+- own a domain, we will use [Google Domains](https://domains.google.com/registrar/) in this guide but any provider will work
+- latest Jenkins X [upgrade](/docs/v3/guides/upgrade)
+
+### Cloud Infrastructure
+First we will configure the cloud infrastructure requirements:
+
+- a GCP Service Account with the `dns.admin` role, see [here](https://cloud.google.com/iam/docs/understanding-roles#dns-roles) for more information
+- a managed cloud dns zone, see [here](https://cloud.google.com/dns/docs/zones) for more information
+
+To satisfy these requirements go to your infrastructure repository (contains Terraform main.tf) and add to your `values.auto.tfvars` the following:
+
+```yaml
+parent_domain = "foo.io"
+```
+
+Most people prefer to use a subdomain for a specific installation rather than purchasing one domain per cluster.  For example in a multi cluster setup you will probably want all using the same parent domain but two clusters using a different subdomain like development.foo.io, staging.foo.io leaving production using just the parent domain foo.io.
+
+To use a subdomain for this cluster add the following configuration:
+
+```yaml
+subdomain     = "bar"
+```
+
+Now apply these changes:
+
+```bash
+git add values.auto.tfvars
+git commit -m 'feat: enable DNS cloud resources'
+git push
+```
+
+```bash
+terraform plan
+terraform apply
+```
+
+You can now see your managed zone in GCP [here](https://console.cloud.google.com/net-services/dns/zones)
+
+
+### Cluster
+
+Next we will configure the cluster requirements:
+
+- Install [external-dns](https://github.com/kubernetes-sigs/external-dns#externaldns) - Kubernetes controller which watches for new Kubernetes Ingress resources and creates A records in Google Cloud DNS which will propogate globally across the internet
+- Install [cert-manager](https://cert-manager.io/docs/) - Kubernetes controller which watches for requests to ask [Let's Encrypt](https://letsencrypt.org/) to issue a new wildcard TLS certificate for your domain and will manage this including renewals
+
+To satisfy these requirements go to your cluster repository (contains helmfile.yaml)
+
+Add external-dns to your clusters helmfile.yaml `releases` section:
+
+```bash
+- chart: bitnami/external-dns
+```
+
+Add cert-manager to your clusters helmfile.yaml `releases` section:
+```bash
+- chart: jetstack/cert-manager
+```
+
+Next we install
+- a cluster wide [Issuer](https://cert-manager.io/docs/concepts/issuer/) which tells cert-manager how to validate you own your domain
+- a namespaced [Certificate](https://cert-manager.io/docs/concepts/certificate/) to request a TLS certificate for services running in the `jx` namespace
+
+```bash
+- chart: jx3/acme
+  name: tls-jx
+  values:
```

Comment:
might want to add something like this to highlight the dependency on cert-manager so we avoid race conditions - chart: jetstack/cert-manager
  name: cert-manager
  namespace: jx
- chart: jx3/acme
  needs:
    - cert-manager
  name: tls-jx
  namespace: jx
  values:
    - issuer:
        enabled: true
        cluster: true

Reply:
thanks!  we might need to think about this one a little further because Jenkins X uses `helmfile template` with an output directory to write the generated kubernetes resources, it then does `kubectl apply` so I'm not totally sure if the 
```
needs:
  - cert-manager/cert-manager
```
will work as expected, there is a an option helmfile template flag
```
--output-dir-template value   go text template for generating the output directory. Default: {{ .OutputDir }}/{{ .State.BaseName }}-{{ .State.AbsPathSHA1 }}-{{ .Release.Name}}
```
which might be useful to see if we can split the CRDs into a different folder from the `ClusterIssuer` and then `kubectl` applies them in the right order?

i.e. 
```
helmfile template --output-dir=cluster --output-dir-template=[something]
```
generates folders:
```
/cluster
- /crds/foo.yaml
- /other cluster resources/bar.yaml
```
then 
```
kubectl apply -f cluster
```
first applies the CRDs and next the other cluster resources?

Reply:
Think you're right! I think the two options are the one  you have suggested where we parse CRDs and modify their path so they are applied first. The only other one which feels like a worse  option is we support nested helmfiles and honour this structure in our `cluster` and `namespace` folders. I can't really think of much of a use case for  nested state  unless our chart list gets very large so for now probably your suggestion of separating and applying  CRDs before other resources makes the most sense. Sounds like an issue against `jx gitops` which I'll open later